### PR TITLE
Implement geo-traits for reading from WKT

### PR DESCRIPTION
**Pull request by kylebarron**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [ ] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
- [x] I ran cargo fmt
---

Ref https://github.com/georust/geo/pull/1157

### CODE REVIEW

File: src/types/coord.rs [hunk 1]
```diff
@@ -96,6 +98,100 @@ where
...
+
+impl<T: WktNum> PointTrait for &Coord<T> {
+    type T = T;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        match (self.z.is_some(), self.m.is_some()) {
+            (true, true) => geo_traits::Dimensions::XYZM,
+            (true, false) => geo_traits::Dimensions::XYZ,
+            (false, true) => geo_traits::Dimensions::XYM,
+            (false, false) => geo_traits::Dimensions::XYM,
```
Comment (michaelkirk) [inline src/types/coord.rs:156]:
```suggestion
            (false, false) => geo_traits::Dimensions::XY,
```

File: src/types/geometrycollection.rs [hunk 2]
```diff
@@ -84,6 +86,27 @@ where
     }
 }
 
+impl<T: WktNum> GeometryCollectionTrait for GeometryCollection<T> {
+    type T = T;
+    type GeometryType<'a> = &'a Wkt<T> where Self: 'a;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        if self.0.is_empty() {
```
Comment (michaelkirk) [inline src/types/geometrycollection.rs:94]: Can this be parsed from the input?

`GEOMETRYCOLLECTION ZM EMPTY`

That might be a larger change though - requiring explicit tracking of the dimensionality of the geometry.

Comment (kylebarron) [reply src/types/geometrycollection.rs:94]: I didn't recall that empty geometries could also have a dimension. Does simple features allow that?

Comment (michaelkirk) [reply src/types/geometrycollection.rs:94]: I guess I'm not sure. I just assumed that this was valid:

`GEOMETRYCOLLECTION Z(POLYGON Z EMPTY, POINT Z(1 2 3))`

Maybe it's not?

Comment (kylebarron) [reply src/types/geometrycollection.rs:94]: Yeah you're right, it is valid: 
image
(I assume shapely/geos implement Simple Features accurately)

Comment (kylebarron) [reply src/types/geometrycollection.rs:94]: I'd say that yes, long-term we should figure out a way to maintain the dimension for empty WKT geometries.

File: src/types/polygon.rs [hunk 3]
```diff
@@ -70,6 +72,58 @@ where
     }
 }
 
+impl<T: WktNum> PolygonTrait for Polygon<T> {
+    type T = T;
+    type RingType<'a> = &'a LineString<T> where Self: 'a;
+
+    fn dim(&self) -> geo_traits::Dimensions {
+        // TODO: infer dimension from empty WKT
```
Comment (michaelkirk) [inline src/types/polygon.rs:80]: Can you please make an issue describing the work to be done here and any context that we might not remember in a hundred years when someone actually does the work?

File: src/types/polygon.rs [hunk 4]
```diff
@@ -70,6 +72,58 @@ where
...
+    fn exterior(&self) -> Option<Self::RingType<'_>> {
+        self.0.first()
+    }
+
+    fn num_interiors(&self) -> usize {
+        self.0.len().saturating_sub(1)
+    }
+
+    unsafe fn interior_unchecked(&self, i: usize) -> Self::RingType<'_> {
+        &self.0[i + 1]
```
Comment (michaelkirk) [inline src/types/polygon.rs:123]: Based on my latest understanding after https://github.com/georust/geojson/pull/245/files#r1821255299,

Other than the `CoordTrait.nth_unchecked`, shouldn't these `unchecked` methods be using https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get_unchecked like https://github.com/georust/geojson/pull/245?

**Review by michaelkirk — CHANGES_REQUESTED** [review]
I'm going to make a PR into geo-traits with the unchecked/or_panic variants so we can stop having the same conversation in every geo-traits repo 🤣

**Comment by michaelkirk** [comment]
(other than the `unchecked` stuff, this looks good!)

File: src/types/polygon.rs (hunk 3 above)
Comment (kylebarron) [reply src/types/polygon.rs:80]: Made https://github.com/georust/wkt/issues/125

**Comment by michaelkirk** [comment]
Let's release geo-traits-0.2.0 first so we can avoid one sem-ver break.

**Comment by kylebarron** [comment]
I updated this to geo-traits 0.2. Can I get a ✅ ?

**Review by michaelkirk — APPROVED** [review]

**Comment by kylebarron** [comment]
Thanks! 🚀

**Pull request merged**